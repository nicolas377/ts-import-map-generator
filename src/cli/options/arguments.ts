// This is where the code that tells the parser what arguments exist, and how to parse and validate them goes.
// This is the only place where the parser should need to be modified to add new arguments.

export type ArgumentType = string | boolean | number;

type StringFromArgumentType<T extends ArgumentType> = T extends string
  ? "string"
  : T extends boolean
  ? "boolean"
  : T extends number
  ? "number"
  : never;

interface BaseArgumentData<T extends ArgumentType> {
  id: ArgumentKind;
  name: string;
  description: string;
  required: boolean;
  type: StringFromArgumentType<T>;
  defaultValue: T;
  singleDashNames: string[];
  doubleDashNames: string[];
  dataGetter: (value: string) => T | Error;
  validator: (value: T) => boolean;
}

interface StringArgumentType extends BaseArgumentData<string> {
  type: "string";
  defaultValue: string;
  singleDashNames: [];
  dataGetter: (value: string) => string | Error;
  validator: (value: string) => boolean;
}

interface BooleanArgumentType extends BaseArgumentData<boolean> {
  type: "boolean";
  defaultValue: boolean;
  singleDashNames: string[];
  dataGetter: (value: string) => boolean | Error;
  validator: (value: boolean) => boolean;
}

interface NumberArgumentType extends BaseArgumentData<number> {
  type: "number";
  defaultValue: number;
  singleDashNames: [];
  dataGetter: (value: string) => number | Error;
  validator: (value: number) => boolean;
}

type ArgumentData =
  | StringArgumentType
  | BooleanArgumentType
  | NumberArgumentType;

export const idToDataMap = new Map<ArgumentKind, ArgumentData>();
export const nameToIdMap = new Map<
  string,
  [singleOrDoubleDash: "single" | "double", kind: ArgumentKind]
>();
export const requiredArgumentIds: ArgumentKind[] = [];

function createArgument(data: ArgumentData): void {
  const { id, singleDashNames, doubleDashNames, required } = data;

  idToDataMap.set(id, data);
  if (required) {
    requiredArgumentIds.push(id);
  }
  for (const name of singleDashNames) {
    nameToIdMap.set(name, ["single", id]);
  }
  for (const name of doubleDashNames) {
    nameToIdMap.set(name, ["double", id]);
  }
}

export const missingArgumentSymbol = Symbol.for("missing");

function numberDataGetter(value: string): number {
  return Number.parseInt(value, 10);
}

function booleanDataGetter(value: string): boolean | Error {
  if (value === "true") {
    return true;
  }
  if (value === "false") {
    return false;
  }
  return new Error(`Invalid boolean value: ${value}`);
}

function noopValidator() {
  return true;
}

// This is where the code that tells the parser what arguments exist, and how to parse and validate them goes.

export const enum ArgumentKind {
  Help,
  Version,
  Entrypoint,
  GraphMaxDepth,
  IgnoreFiles,
}

createArgument({
  id: ArgumentKind.Help,
  name: "help",
  description: "Show this help message.",
  required: false,
  type: "boolean",
  defaultValue: false,
  singleDashNames: ["h"],
  doubleDashNames: ["help"],
  dataGetter: booleanDataGetter,
  validator: noopValidator,
});

createArgument({
  id: ArgumentKind.Version,
  name: "version",
  description: "Show the version of the import map generator.",
  required: false,
  type: "boolean",
  defaultValue: false,
  singleDashNames: ["v"],
  doubleDashNames: ["version"],
  dataGetter: booleanDataGetter,
  validator: noopValidator,
});

createArgument({
  id: ArgumentKind.Entrypoint,
  name: "entrypoint",
  description: "The entrypoint to generate an import map from.",
  required: true,
  type: "string",
  defaultValue: missingArgumentSymbol as unknown as string,
  singleDashNames: [],
  doubleDashNames: ["entrypoint"],
  dataGetter: (value) => value,
  validator: (value) => value !== "",
});

createArgument({
  id: ArgumentKind.GraphMaxDepth,
  name: "graph-max-depth",
  description: "The maximum depth of the graph to generate.",
  required: false,
  type: "number",
  defaultValue: 1000,
  singleDashNames: [],
  doubleDashNames: ["graph-max-depth", "max-depth"],
  dataGetter: numberDataGetter,
  validator: (value) => value >= 0,
});

createArgument({
  id: ArgumentKind.IgnoreFiles,
  name: "ignore-files",
  description: "A comma-separated list of globs to ignore when matched.",
  required: false,
  type: "string",
  defaultValue: "",
  singleDashNames: [],
  doubleDashNames: ["ignore-files", "ignore"],
  dataGetter: (value) => value,
  // TODO: Validate that the value is a comma-separated list of valid globs when implementing globs.
  validator: noopValidator,
});
