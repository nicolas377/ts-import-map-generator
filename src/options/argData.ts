// This is where the code that tells the parser what arguments exist, and how to parse and validate them goes.

type ArgumentType = "string" | "boolean" | "number";

type TypeFromArgumentType<T extends ArgumentType> = T extends "string"
  ? string
  : T extends "boolean"
  ? boolean
  : number;

interface ArgumentData<T extends ArgumentType> {
  id: ArgumentKind;
  name: string;
  description: string;
  type: T;
  defaultValue: TypeFromArgumentType<T>;
  singleDashNames: T extends "boolean" ? string[] : [];
  doubleDashNames: string[];
  validator?: (value: TypeFromArgumentType<T>) => boolean;
}

const nameToIdTable: Record<string, ArgumentKind> = {};

function createArgument<T extends ArgumentType>(
  id: ArgumentKind,
  name: string,
  description: string,
  type: T,
  defaultValue: TypeFromArgumentType<T>,
  singleDashNames: T extends "boolean" ? string[] : [],
  doubleDashNames: string[],
  /** For strings, you can use RegExp.prototype.test */
  validator?: (value: TypeFromArgumentType<T>) => boolean
): ArgumentData<T> {
  nameToIdTable[name] = id;

  return {
    id,
    name,
    description,
    type,
    singleDashNames,
    doubleDashNames,
    defaultValue,
    validator,
  };
}

const enum ArgumentKind {}
