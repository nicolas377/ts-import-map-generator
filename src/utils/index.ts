export * from "utils/debug";
export * from "utils/generalHelpers";
export * from "utils/options";
