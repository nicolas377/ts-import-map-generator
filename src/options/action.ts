import { idToDataMap, nameToIdMap } from "./arguments";
import { NodeFlags, parseSyntaxTreeFromArgsString } from "./parser";

type OptionValue = string | boolean | number;

interface ActionToTake {
  name: string;
  value: OptionValue;
  isDefault: boolean;
}

export function takeActionFromCliArgs(argsString: string): void {
  // This is the real worker of the argument parser.
  // It takes the string of arguments, parses it, and then takes the appropriate action based on the parsed arguments.
  // We want everything to be as modular as possible, so we rely on data provided in arguments.ts to do the argument-specific work.
  // The goal of this function is to set everything in programOptions before starting the actual work.

  const syntaxTreeApi = parseSyntaxTreeFromArgsString(argsString);

  const actionsToTake: ActionToTake[] = [];

  for (const treeArgument of syntaxTreeApi.syntaxTree.arguments) {
    // Try to parse the argument's id from its name. If it can't be parsed, then it's an unknown argument and should be ignored.
    const argumentId = nameToIdMap.get(treeArgument.flag.text);

    if (argumentId === undefined) {
      continue;
    }

    // The argument is known, so we can get its data from the map.
    const optionDetails = idToDataMap.get(argumentId)!;
    const { value: argumentValue } = treeArgument;

    // TODO: pick up here
  }

  console.log(actionsToTake);
}
