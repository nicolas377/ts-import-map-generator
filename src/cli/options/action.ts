import {
  ArgumentKind,
  ArgumentType,
  idToDataMap,
  nameToIdMap,
} from "./arguments";
import { parseSyntaxTreeFromArgsString } from "./parser";
import { Debug, programOptions } from "utils";

interface ActionToTake {
  id: ArgumentKind;
  value: ArgumentType;
}

export function takeActionFromCliArgs(argsString: string): void {
  // This is the real worker of the argument parser.
  // It takes the string of arguments, parses it, and then takes the appropriate action based on the parsed arguments.
  // We want everything to be as modular as possible, so we rely on data provided in arguments.ts to do the argument-specific work.
  // The goal of this function is to set everything in programOptions before starting the actual work.

  const syntaxTreeApi = parseSyntaxTreeFromArgsString(argsString);

  const actionsToTake: ActionToTake[] = [];

  for (const treeArgument of syntaxTreeApi.syntaxTree.arguments) {
    const argumentFlagName = treeArgument.flag.text;
    // Try to parse the argument's id from its name. If it can't be parsed, then it's an unknown argument and should be ignored.
    const [singleOrDoubleDash, argumentId] =
      nameToIdMap.get(argumentFlagName) ?? [];
    const optionDetails = idToDataMap.get(argumentId!);
    const { dash: argumentDash, value: argumentValue } = treeArgument;

    if (
      singleOrDoubleDash === undefined ||
      argumentId === undefined ||
      optionDetails === undefined ||
      // The argument is known, but has the wrong dash.
      (singleOrDoubleDash === "single" && argumentDash.singleDash === false) ||
      (singleOrDoubleDash === "double" && argumentDash.doubleDash === false)
    ) {
      // The argument is unknown, so we should ignore it.
      Debug.log.warn(`Unknown argument: ${argumentFlagName}`);
      continue;
    }

    // If the argument is a boolean argument, then it doesn't have a value.
    // If there's no value, then set the value to the default value.
    // If there is a value, then set the value to the parsed value.
    if (optionDetails.type === "boolean") {
      const value =
        argumentValue === undefined
          ? true
          : optionDetails.dataGetter(argumentValue.text);

      if (value instanceof Error) {
        Debug.log.warn(`Invalid value for argument: ${argumentFlagName}`);
        continue;
      }

      if (optionDetails.validator(value)) {
        if (value !== optionDetails.defaultValue)
          addAction(optionDetails.id, value);
      } else {
        Debug.log.warn(`Invalid value for argument: ${argumentFlagName}`);
      }
    } else {
      if (argumentValue === undefined) {
        Debug.log.warn(`Missing value for argument: ${argumentFlagName}`);
        continue;
      }

      // The argument is known and has a value, so we should parse it and take the appropriate action.
      const value = optionDetails.dataGetter(argumentValue.text);

      if (value instanceof Error) {
        Debug.log.warn(`Invalid value for argument: ${argumentFlagName}`);
        continue;
      }

      if (
        (optionDetails.validator as (value: string | number) => boolean)(
          value
        ) &&
        value !== optionDetails.defaultValue
      ) {
        addAction(optionDetails.id, value);
      } else {
        Debug.log.warn(
          `Invalid value ${argumentValue.text} for argument: ${argumentFlagName}`
        );
      }
    }
  }

  for (const { id, value } of actionsToTake) {
    programOptions.setOption(id, value);
  }

  function addAction(id: ArgumentKind, value: ArgumentType): void {
    actionsToTake.push({
      id,
      value,
    });
  }
}
