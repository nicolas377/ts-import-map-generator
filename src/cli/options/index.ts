import { argv } from "node:process";
import { takeActionFromCliArgs } from "./action";

export function initializeOptionsFromCliArgs(): void {
  const argsString: string = argv.slice(2).join(" ");

  takeActionFromCliArgs(argsString);
}
