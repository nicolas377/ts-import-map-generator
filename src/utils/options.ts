import { Debug } from "./debug";
import { keysOfObject } from "./generalHelpers";
import { ArgumentKind, ArgumentType, idToDataMap } from "cli/options/arguments";

const argumentKindToOptionNameMap: Readonly<
  Record<ArgumentKind, keyof IOptions>
> = {
  [ArgumentKind.Help]: "printHelpAndExit",
  [ArgumentKind.Version]: "printVersionAndExit",
  [ArgumentKind.Entrypoint]: "entrypointLocation",
  [ArgumentKind.GraphMaxDepth]: "graphMaxDepth",
  [ArgumentKind.IgnoreFiles]: "ignoreFiles",
};

interface IOptions {
  getOption(kind: ArgumentKind): ArgumentType;
  setOption(kind: ArgumentKind, value: ArgumentType): void;
  printHelpAndExit: boolean;
  printVersionAndExit: boolean;
  entrypointLocation: string;
  graphMaxDepth: number;
  ignoreFiles: string;
}

class OptionsClass {
  private allOptions = keysOfObject(argumentKindToOptionNameMap);

  private options: Record<ArgumentKind, ArgumentType>;

  public getOption(kind: ArgumentKind): ArgumentType {
    return this.options[kind];
  }

  public setOption(kind: ArgumentKind, value: ArgumentType): void {
    this.options[kind] = value;
  }

  constructor() {
    // kind should be one of all of the options, so this validates that at compile-time.
    this.options = this.allOptions.reduce((acc, kind: ArgumentKind) => {
      const { defaultValue } = idToDataMap.get(+kind) ?? {};

      Debug.assertIsDefined(
        defaultValue,
        `No default value for argument kind ${kind}`
      );

      acc[kind] = defaultValue;
      return acc;
    }, {} as Record<ArgumentKind, ArgumentType>);

    for (const kind of this.allOptions) {
      const optionData = idToDataMap.get(+kind);

      Debug.assertIsDefined(
        optionData,
        `No option data for argument kind ${kind}`
      );

      Object.defineProperty(this, argumentKindToOptionNameMap[kind], {
        configurable: false,
        enumerable: true,
        get: () => this.options[kind],
        set: (value: ArgumentType) => {
          // TODO: when TS supports merging function types, this assertion can be removed.
          (optionData.validator as (value: ArgumentType) => boolean)(value);
          this.options[kind] = value;
        },
      });
    }
  }
}

export const programOptions = new OptionsClass() as unknown as IOptions;
