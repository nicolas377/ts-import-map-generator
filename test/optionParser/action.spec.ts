// Integration tests for the option parser

import { objectWithoutProperties } from "../helpers";
import { takeActionFromCliArgs } from "cli/options/action";
import { programOptions } from "utils/options";

describe("Option parser integration", () => {
  const emptyProgramOptions = { ...programOptions };

  afterEach(() => {
    Object.assign(programOptions, emptyProgramOptions);
  });

  const testCases: [
    commandLineArgv: string,
    optionClassKey: keyof typeof programOptions,
    expectedValue: unknown
  ][] = [
    ["--help", "printHelpAndExit", true],
    ["--version", "printVersionAndExit", true],
    ["--entrypoint=./src/index.ts", "entrypointLocation", "./src/index.ts"],
    ['--entrypoint="./src/index.ts"', "entrypointLocation", "./src/index.ts"],
    ["--graph-max-depth=900", "graphMaxDepth", 900],
    ["--ignore-files=**/*.test.ts", "ignoreFiles", "**/*.test.ts"],
    ['--ignore-files="**/*.test.ts"', "ignoreFiles", "**/*.test.ts"],
  ];

  test.each(testCases)(
    "should parse the %p option correctly",
    (commandLineArgv, optionClassKey, expectedValue) => {
      takeActionFromCliArgs(commandLineArgv);

      expect(programOptions[optionClassKey]).toBe(expectedValue);
    }
  );

  test.each(testCases)(
    "should not touch anything but the %p option",
    (commandLineArgv, optionClassKey) => {
      takeActionFromCliArgs(commandLineArgv);

      expect(objectWithoutProperties(programOptions, [optionClassKey])).toEqual(
        objectWithoutProperties(emptyProgramOptions, [optionClassKey])
      );
    }
  );

  test.each(testCases)("snapshot of the %p option", (commandLineArgv) => {
    takeActionFromCliArgs(commandLineArgv);

    expect(programOptions).toMatchSnapshot();
  });

  test("should not touch anything if no options are passed", () => {
    takeActionFromCliArgs("");

    expect(programOptions).toEqual(emptyProgramOptions);
  });

  test("empty snapshot", () => {
    expect(emptyProgramOptions).toMatchSnapshot();
  });
});
