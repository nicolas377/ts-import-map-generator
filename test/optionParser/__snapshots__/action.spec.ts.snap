// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option parser integration empty snapshot 1`] = `
{
  "allOptions": [
    "0",
    "1",
    "2",
    "3",
    "4",
  ],
  "entrypointLocation": Symbol(missing),
  "graphMaxDepth": 1000,
  "ignoreFiles": "",
  "options": {
    "0": false,
    "1": false,
    "2": Symbol(missing),
    "3": 1000,
    "4": "",
  },
  "printHelpAndExit": false,
  "printVersionAndExit": false,
}
`;

exports[`Option parser integration snapshot of the "--entrypoint=./src/index.ts" option 1`] = `
OptionsClass {
  "allOptions": [
    "0",
    "1",
    "2",
    "3",
    "4",
  ],
  "entrypointLocation": ".",
  "graphMaxDepth": 1000,
  "ignoreFiles": "",
  "options": {
    "0": false,
    "1": false,
    "2": ".",
    "3": 1000,
    "4": "",
  },
  "printHelpAndExit": false,
  "printVersionAndExit": false,
}
`;

exports[`Option parser integration snapshot of the "--entrypoint=\\"./src/index.ts\\"" option 1`] = `
OptionsClass {
  "allOptions": [
    "0",
    "1",
    "2",
    "3",
    "4",
  ],
  "entrypointLocation": """,
  "graphMaxDepth": 1000,
  "ignoreFiles": "",
  "options": {
    "0": false,
    "1": false,
    "2": """,
    "3": 1000,
    "4": "",
  },
  "printHelpAndExit": false,
  "printVersionAndExit": false,
}
`;

exports[`Option parser integration snapshot of the "--graph-max-depth=900" option 1`] = `
OptionsClass {
  "allOptions": [
    "0",
    "1",
    "2",
    "3",
    "4",
  ],
  "entrypointLocation": Symbol(missing),
  "graphMaxDepth": 900,
  "ignoreFiles": "",
  "options": {
    "0": false,
    "1": false,
    "2": Symbol(missing),
    "3": 900,
    "4": "",
  },
  "printHelpAndExit": false,
  "printVersionAndExit": false,
}
`;

exports[`Option parser integration snapshot of the "--help" option 1`] = `
OptionsClass {
  "allOptions": [
    "0",
    "1",
    "2",
    "3",
    "4",
  ],
  "entrypointLocation": Symbol(missing),
  "graphMaxDepth": 1000,
  "ignoreFiles": "",
  "options": {
    "0": true,
    "1": false,
    "2": Symbol(missing),
    "3": 1000,
    "4": "",
  },
  "printHelpAndExit": true,
  "printVersionAndExit": false,
}
`;

exports[`Option parser integration snapshot of the "--ignore-files=**/*.test.ts" option 1`] = `
OptionsClass {
  "allOptions": [
    "0",
    "1",
    "2",
    "3",
    "4",
  ],
  "entrypointLocation": Symbol(missing),
  "graphMaxDepth": 1000,
  "ignoreFiles": "*",
  "options": {
    "0": false,
    "1": false,
    "2": Symbol(missing),
    "3": 1000,
    "4": "*",
  },
  "printHelpAndExit": false,
  "printVersionAndExit": false,
}
`;

exports[`Option parser integration snapshot of the "--ignore-files=\\"**/*.test.ts\\"" option 1`] = `
OptionsClass {
  "allOptions": [
    "0",
    "1",
    "2",
    "3",
    "4",
  ],
  "entrypointLocation": Symbol(missing),
  "graphMaxDepth": 1000,
  "ignoreFiles": """,
  "options": {
    "0": false,
    "1": false,
    "2": Symbol(missing),
    "3": 1000,
    "4": """,
  },
  "printHelpAndExit": false,
  "printVersionAndExit": false,
}
`;

exports[`Option parser integration snapshot of the "--version" option 1`] = `
OptionsClass {
  "allOptions": [
    "0",
    "1",
    "2",
    "3",
    "4",
  ],
  "entrypointLocation": Symbol(missing),
  "graphMaxDepth": 1000,
  "ignoreFiles": "",
  "options": {
    "0": false,
    "1": true,
    "2": Symbol(missing),
    "3": 1000,
    "4": "",
  },
  "printHelpAndExit": false,
  "printVersionAndExit": true,
}
`;
